### ISSUE TITLE
Test suite crashes on Xcode 13.3 / Swift 5.6

### ISSUE THREAD

**Issue by nighthawk** [nicklockwood/Expression#37]
Running this repo's test suite on macOS 12.3 with Xcode 13.3 and Swift 5.6 crashes on my M1 MacBook Air with an `EXC_BAD_ACCESS`.

This happens in tests such as `testTooFewArgumentsForCustomFunctionWithAdvancedInitializer` and a number of other ones which test a mismatching number of arguments.

```
#0	0x000000016fdfcd98 in 0x16fdfcd98 ()
#1	0x00000001078bc278 in thunk for @escaping @callee_guaranteed (@in_guaranteed [Double]) -> (@out Double, @error @owned Error) ()
#2	0x00000001078f92e0 in partial apply for thunk for @escaping @callee_guaranteed (@in_guaranteed [Double]) -> (@out Double, @error @owned Error) ()
#3	0x00000001078dfba8 in Subexpression.optimized(withImpureSymbols:pureSymbols:) at /Users/adrian/Development/forks/Expression/Sources/Expression.swift:892
#4	0x00000001078de7ac in Expression.init(_:impureSymbols:pureSymbols:) at /Users/adrian/Development/forks/Expression/Sources/Expression.swift:351
#5	0x00000001078de6e0 in Expression.__allocating_init(_:impureSymbols:pureSymbols:) ()
#6	0x00000001078dfe20 in Expression.__allocating_init(_:pureSymbols:) at /Users/adrian/Development/forks/Expression/Sources/Expression.swift:372
#7	0x000000010788b350 in ExpressionTests.testTooFewArgumentsForCustomFunctionWithAdvancedInitializer() at /Users/adrian/Development/forks/Expression/Tests/ExpressionTests.swift:934
```

I couldn't yet narrow down what might be causing it and if it's an issue introduced with Swift 5.6 or with this library, or what to do about it. It happens both when running through Xcode or with `swift test` from the command line. Any advise appreciated!

**Comment by nighthawk** [comment]
Screenshot of the crash in Xcode, in case it helps.

<img width="830" alt="Screen Shot 2022-03-24 at 14 31 31" src="https://user-images.githubusercontent.com/22646/159836639-2a7f0b24-1b9a-4011-9d88-662d6b7fe2a0.png">

**Comment by nicklockwood** [comment]
Looks like a Swift 5.6 regression. I'll investigate.

**Comment by nicklockwood** [comment]
@nighthawk is the crash affecting your project, or just the test failure? If it's only the latter I can just disable the test for now.

**Comment by nighthawk** [comment]
As far as I can tell, it doesn't seem to be impacting my project; just running the tests crashed.

I commented out these tests temporarily:

- `testTooFewArgumentsWithAdvancedInitializer`
- `testTooFewArgumentsForCustomFunctionWithAdvancedInitializer` 
- `testTooManyArgumentsWithAdvancedInitializer`
- `testUnknownSymbolWithAdvancedInitializer`

**Comment by bhimsenp** [comment]
@nicklockwood .. Seems following line is causing the issue (Expression.swift: line 354)
```
if let fn = pureSymbols($0) ?? Expression.mathSymbols[$0] ?? Expression.boolSymbols[$0] {
    return fn
}
```
Haven't figured out the exact cause yet, but looks like there is an issue where there is cascaded nil-coalescing operation.
In one of the failing test, the `fn` evaluates to non-nil even when all 3 expressions are nil. Consequently, when it tries to execute line 892 (where crash is happening), it tries to execute the nil function and that's why the `EXEC_BAD_ACCESS`.

Adding brackets to first two correctly evaluated fn to be nil solves the crash.
```
if let fn = (pureSymbols($0) ?? Expression.mathSymbols[$0]) ?? Expression.boolSymbols[$0] {
    return fn
}
```
You can add this workaround temporarily.

**Comment by haikusw** [comment]
Reproduced this here, macOS 12.2.1 (21D62) / Xcode 13.3, 13" M1 MBP.

**Comment by haikusw** [comment]
`fn	()	0x000000011b8525bc Expression`reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed Swift.Array<Swift.Double>) -> (@out Swift.Double, @error @owned Swift.Error) to @escaping @callee_guaranteed (@guaranteed Swift.Array<Swift.Double>) -> (@unowned Swift.Double, @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".95" at <compiler-generated>


(lldb) po argValues
▿ 1 element
  - 0 : 4.0
 `

**Comment by haikusw** [comment]
`    /// Alternative constructor for advanced usage
    /// Allows for dynamic symbol lookup or generation without any performance overhead
    /// Note that both math and boolean symbols are enabled by default - to disable them
    /// return `{ _ in throw Expression.Error.undefinedSymbol(symbol) }` from your lookup function
    public init(
        _ expression: ParsedExpression,
        impureSymbols: (Symbol) -> SymbolEvaluator?,
        pureSymbols: (Symbol) -> SymbolEvaluator? = { _ in nil }
    ) {
        root = expression.root.optimized(
            withImpureSymbols: impureSymbols,
            pureSymbols: {
                if let fn = pureSymbols($0) ?? Expression.mathSymbols[$0] ?? Expression.boolSymbols[$0] {
                    return fn
                }
`

This `if let fn =` doesn't seem to be doing the right thing.

`pureSymbols($0)` seems to produce nil and so do `mathSymbols[$0]` and `boolSymbols[$0]` but the `return fn` is getting executed even though it seems like the `if let` should have failed.

**Comment by haikusw** [comment]
Disassembly is too long to put here, but it's [here](https://gist.github.com/haikusw/b53652edb50f3336cec2e9a2b519c262).

**Comment by haikusw** [comment]
Confirming that this test passes in Xcode Version 13.2.1 (13C100)

**Comment by haikusw** [comment]
work around - put the first expression in parens to help the compiler:

`if let fn = (pureSymbols($0) ?? Expression.mathSymbols[$0]) ?? Expression.boolSymbols[$0] {`

Test passes in Xcode 13.3 if I do that.

**Comment by nicklockwood** [comment]
@bhimsenp thanks, that solved it!

**Comment by nicklockwood** [comment]
@nighthawk fixed in 0.13.6
